File: CSOC-Week0/Challenge-2/To Do List progarm/Single User/Python/todolist.py
import ast
import os

# check if file exists
def check_file():
    if not os.path.exists('tasklist.txt'):
        print("No task list exists...")
        inp = input("Enter 'quit' to exit the program or any other letter to create a empty list: ")
        if inp == 'quit':
            exit()
        else:
            create_empty_task_list() #create emply file if not exits
    return 1

#list all the task
def list_task():
    with open("tasklist.txt", "r") as task_file:
        data = task_file.read().splitlines()
        num = 1
        for i in data:
            print("{}. {}".format(num, ast.literal_eval(i)["task"]))
            num = num + 1
    initialize()

#adding task
def update_task_list():
    with open("tasklist.txt", "r") as task_file:
        data = task_file.read().splitlines()
    with open("tasklist.txt", "a") as task_file:
        try:
            last_UID = int(ast.literal_eval(data[-1])["UID"])
        except IndexError:
            last_UID = 0
        data_block = {"UID": str(last_UID + 1), "task": input("Enter your task:")}
        task_file.write(str(data_block))
        task_file.write("\n")
    initialize()

# takeing user input in menu
def take_input():
    try:
        task = int(input("Enter Choice (enter '0' for options)>>>"))
    except ValueError:
        take_input()
        return
    while task not in range(0, 6):
        task = int(input("Enter Choice (enter '0' for options)"))
    return task

# take input to delete the task
def take_del_input(num):
    try:
        task = int(input(">>>"))
    except ValueError:
        take_del_input(num)
        return
    while task not in range(1, num):
        task = int(input(">>>"))
    return task

# deleting task
def delete_task():
    with open("tasklist.txt", "r+") as task_file:
        data = task_file.read().splitlines()
        num = 1
        for i in data:
            print("{}. {}".format(num, ast.literal_eval(i)["task"]))
            num = num + 1
        print("Enter which task you want to delete?")
        task = take_del_input(num)
        data.pop(task - 1)
    with open("tasklist.txt", "w") as task_file:
        for i in data:
            task_file.write(i)
            task_file.write("\n")
    print("Deleted...")
    initialize()

# confirmation to clr task list
def confirmation():
    try:
        inp = input("Do you really want to clr task list (YES/NO)").lower()
        if inp == 'yes':
            return 1
        elif inp == 'no':
            return 0
        else:
            confirmation()
    except Exception:
        confirmation()

# create empty task file
def create_empty_task_list():
    with open("tasklist.txt", "w") as task_file:
        pass
    initialize()

# clear task file
def clear_task_list():
    confirm = confirmation()
    if confirm == 1:
        create_empty_task_list()
    else:
        initialize()

# help option
def task_list_help():
    print("1: List Your Tasks\n2: Update Task List\n3: Delete Task\n4: Clear task List\n5:Exit\n")
    initialize()

# intitializing the list
def initialize():
    task = take_input()
    if task == 0:
        task_list_help()
    elif task == 1:
        list_task()
    elif task == 2:
        update_task_list()
    elif task == 3:
        delete_task()
    elif task == 4:
        clear_task_list()
    else:
        exit(0)

# program start
if __name__ == '__main__':
    print("Welcome to To-Do List:")
    if check_file():
        task_list_help()
        initialize()

File: CSOC-Week0/Challenge-2/To Do List progarm/Single User/C/todolist.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define FILENAME "tasklistc.txt"
#define MAX_TASK_LENGTH 256

// task structure
typedef struct {
    int UID;
    char task[MAX_TASK_LENGTH];
} Task;

// Global variables for task list and task count
Task *task_list = NULL;
int task_count = 0;

// Function 
void create_empty_task_list();
void initialize();
void load_tasks();
void save_tasks();

//  check if the task file exists
void check_file() {
    FILE *file = fopen(FILENAME, "r");
    if (!file) {
        // If file does not exist, prompt the user to create a new task list or exit
        printf("No task list exists...\n");
        char inp[MAX_TASK_LENGTH];
        printf("Enter 'quit' to exit the program or any other letter to create an empty list: ");
        fgets(inp, sizeof(inp), stdin);
        inp[strcspn(inp, "\n")] = 0;   
        if (strcmp(inp, "quit") == 0) {
            exit(0);
        } else {
            create_empty_task_list();
        }
    } else {
        fclose(file);
        load_tasks(); // Load existing tasks if file exists
    }
}

// load task
void load_tasks() {
    FILE *file = fopen(FILENAME, "r");
    if (!file) {
        perror("Failed to open task list");
        return;
    }
    task_count = 0;
    char line[MAX_TASK_LENGTH];
    while (fgets(line, sizeof(line), file)) {
        task_list = realloc(task_list, (task_count + 1) * sizeof(Task));
        sscanf(line, "{UID: %d, task: %[^\n]}", &task_list[task_count].UID, task_list[task_count].task);
        task_count++;
    }
    fclose(file);
}

// save task
void save_tasks() {
    FILE *file = fopen(FILENAME, "w");
    if (!file) {
        perror("Failed to save task list");
        return;
    }
    for (int i = 0; i < task_count; i++) {
        fprintf(file, "{UID: %d, task: %s}\n", task_list[i].UID, task_list[i].task);
    }
    fclose(file);
}

//  list all 
void list_task() {
    for (int i = 0; i < task_count; i++) {
        printf("%d. %s\n", i + 1, task_list[i].task);
    }
    initialize();
}

//  enter new task
void update_task_list() {
    task_list = realloc(task_list, (task_count + 1) * sizeof(Task));
    task_list[task_count].UID = (task_count > 0) ? task_list[task_count - 1].UID + 1 : 1;
    printf("Enter your task: ");
    fgets(task_list[task_count].task, sizeof(task_list[task_count].task), stdin);
    task_list[task_count].task[strcspn(task_list[task_count].task, "\n")] = 0; 
    task_count++;
    save_tasks();
    initialize();
}

// taking user input 
int take_input() {
    int task;
    printf("Enter Choice (enter '0' for options)>>> ");
    while (scanf("%d", &task) != 1) {
        while (getchar() != '\n');  
        printf("Enter Choice (enter '0' for options)>>> ");
    }
    while (getchar() != '\n'); 
    return task;
}

// take input to delete task
int take_del_input(int num) {
    int task;
    printf(">>> ");
    while (scanf("%d", &task) != 1 || task < 1 || task >= num) {
        while (getchar() != '\n');  
        printf(">>> ");
    }
    while (getchar() != '\n');
    return task;
}

//  delete task
void delete_task() {
    for (int i = 0; i < task_count; i++) {
        printf("%d. %s\n", i + 1, task_list[i].task);
    }
    printf("Enter which task you want to delete? ");
    int task_to_delete = take_del_input(task_count + 1);
    for (int i = task_to_delete - 1; i < task_count - 1; i++) {
        task_list[i] = task_list[i + 1];
    }
    task_list = realloc(task_list, (task_count - 1) * sizeof(Task));
    task_count--;
    save_tasks();
    printf("Deleted...\n");
    initialize();
}

// confirm clear task
int confirmation() {
    char inp[MAX_TASK_LENGTH];
    printf("Do you really want to clear task list (YES/NO) ");
    fgets(inp, sizeof(inp), stdin);
    inp[strcspn(inp, "\n")] = 0;   
    if (strcmp(inp, "yes") == 0) {
        return 1;
    } else if (strcmp(inp, "no") == 0) {
        return 0;
    } else {
        return confirmation();
    }
}

// create empty task list
void create_empty_task_list() {
    free(task_list);
    task_list = NULL;
    task_count = 0;
    save_tasks();
    initialize();
}

// clear task list
void clear_task_list() {
    if (confirmation() == 1) {
        create_empty_task_list();
    } else {
        initialize();
    }
}

// help
void task_list_help() {
    printf("1: List Your Tasks\n2: Update Task List\n3: Delete Task\n4: Clear task List\n5: Exit\n");
    initialize();
}

// initialize 
void initialize() {
    int task = take_input();
    if (task == 0) {
        task_list_help();
    } else if (task == 1) {
        list_task();
    } else if (task == 2) {
        update_task_list();
    } else if (task == 3) {
        delete_task();
    } else if (task == 4) {
        clear_task_list();
    } else {
        exit(0);
    }
}

// Main function
int main() {
    printf("Welcome to To-Do List:\n");
    check_file();
    task_list_help();
    initialize();
    return 0;
}
